src/main.c

```diff
@@ -1136,32 +1138,11 @@ static int open_listener(int domain, int type, int protocol, struct sockaddr *ad
 {
     int fd;
 
-    if ((fd = socket(domain, type, protocol)) == -1)
+    fd = h2o_priv_open_listener(domain, type, protocol,
+      (struct sockaddr_storage *)addr, addrlen, reuseport);
+    if (fd == -1) {
         goto Error;
-    set_cloexec(fd);
-    /* if the socket is TCP, set SO_REUSEADDR flag to avoid TIME_WAIT after shutdown */
-    if (type == SOCK_STREAM) {
-        int flag = 1;
-        if (setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &flag, sizeof(flag)) != 0)
-            goto Error;
-    }
-#ifdef IPV6_V6ONLY
-    /* set IPv6only */
-    if (domain == AF_INET6) {
-        int flag = 1;
-        if (setsockopt(fd, IPPROTO_IPV6, IPV6_V6ONLY, &flag, sizeof(flag)) != 0)
-            goto Error;
     }
-#endif
-    if (reuseport) {
-#if H2O_HTTP3_USE_REUSEPORT
-        int flag = 1;
-        if (setsockopt(fd, SOL_SOCKET, H2O_SO_REUSEPORT, &flag, sizeof(flag)) != 0)
-            fprintf(stderr, "[warning] setsockopt(SO_REUSEPORT) failed:%s\n", strerror(errno));
-#endif
-    }
-    if (bind(fd, addr, addrlen) != 0)
-        goto Error;
```

Comment:
That should be fine. We are doing something similar with connections. We are performing multiple syscalls in the privileged process, and we are using iovecs to convey the return values, errno values so they can be independently checked. This was done to reduce the number of context switches which may not be necessary here. I can clean this up.

Reply:
Also I don't believe privsep is strictly about wrapping syscalls, in many cases it's not when you consider things like getting new file descriptors for log rotation. In these cases, I've often seen privileges called GET_NEW_LOGGING_FD which is completely decoupled from the underlying syscall, and in some cases, a preferred way to do it. Since syscall wrapping will lead to more work in terms of validating the socket type, path etc that is actually being requested